### PR TITLE
Close feed commands

### PR DESCRIPTION
**Pull request by RodrigoAD**

### CODE REVIEW

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/withdrawFunds.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+const withdrawFundsInstruction: AbstractInstruction<CommandInput, ContractInput> = {
+  instruction: {
+    category: CATEGORIES.OCR,
+    contract: 'ocr2',
+    function: 'withdraw_funds',
+  },
+  examples: [
+    'yarn gauntlet ocr2:withdraw_funds --network=<NETWORK> --all <CONTRACT_ADDRESS>',
+    'yarn gauntlet ocr2:withdraw_funds --network=<NETWORK> --amount=<AMOUNT> <CONTRACT_ADDRESS>',
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/withdrawFunds.ts:68]: Let's document how to withdraw funds for node ops. @kylesmartin could help out here as he's currently owning the documentation story withdrawal is part of.

Comment (RodrigoAD) [reply packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/withdrawFunds.ts:68]: I'll add the command for that too. This one is for withdrawing the total feed funds.

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/acceptProposal.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+type ContractInput = {
+  id: string
+  digest: string
+}
+
+const makeCommandInput = async (flags: any, args: string[]): Promise<CommandInput> => {
+  if (flags.input) return flags.input as CommandInput
+  return {
+    proposalId: flags.proposalId || flags.configProposal, // --configProposal alias requested by eng ops
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/acceptProposal.ts:21]: Let's allow the same params names as contract expects:

```suggestion
    proposalId: flags.id || flags.proposalId || flags.configProposal, // --configProposal alias requested by eng ops
```

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/acceptProposal.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+const instruction: AbstractInstruction<CommandInput, ContractInput> = {
+  examples: [
+    'yarn gauntlet ocr2:accept_proposal:close --network=<NETWORK> --configProposal=<PROPOSAL_ID> --digest=<DIGEST> <CONTRACT_ADDRESS>',
+  ],
+  instruction: {
+    category: CATEGORIES.OCR,
+    contract: 'ocr2',
+    function: 'accept_proposal',
+    subInstruction: 'close',
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/acceptProposal.ts:78]: IMO this is confusing as there is no concept of a `subInstruction`, but this is just a command suffix, right? 

The `:close` is actually a type of `accept_proposal`? 
Why isn't it using the underlying `ocr2:accept_proposal` implementation, but instead reimplementing it from scratch as a special command?

Comment (krebernisak) [reply packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/acceptProposal.ts:78]: Consider a general `cw20:transfer` command and a use case where we want to implement a `cw20:transfer:bob` command, which is a special type of transfer to Bob.

The framework should allow us to easily compose commands like this, by just adding a few lines of code to prepare the data and forward the call to the underlying command.

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/proposeConfig.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+const makeCommandInput = async (flags: any, args: string[]): Promise<CommandInput> => {
+  if (flags.input) return flags.input as CommandInput
+  return {
+    proposalId: flags.configProposal,
+  }
+}
+
+const makeContractInput = async (input: CommandInput): Promise<ContractInput> => {
+  const hexToBase64 = (s) => Buffer.from(s, 'hex').toString('base64')
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/proposeConfig.ts:20]: I've seen an fn like this defined in-line in multiple places. Consider extracting it somewhere.

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/proposeOffchainConfig.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
+import { RDD } from '@chainlink/gauntlet-terra'
+import { AbstractInstruction, instructionToCommand, BeforeExecute } from '../../../abstract/executionWrapper'
+import { CATEGORIES } from '../../../../lib/constants'
+import { logger, prompt } from '@chainlink/gauntlet-core/dist/utils'
+import { ContractInput } from '../proposeOffchainConfig'
+
+export const EMPTY_CONFIG = Buffer.from([1, 2]).toString('base64')
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/proposeOffchainConfig.ts:7]: Why `Buffer.from([1, 2])`?

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/proposeConfig.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+const makeInput = (flags, args): CommandInput => {
+  if (flags.input) return flags.input as CommandInput
+  return {
+    proposalId: flags.configProposal,
+  }
+}
+
+const makeInnerCommandInput = (input: CommandInput): ProposeConfigInput => {
+  const randomAcc = () => new MnemonicKey().publicKey?.address()!
```
Comment (RodrigoAD) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/proposeConfig.ts:15]: - [ ] Generate 4 vanity addresses that represent empty transmitters/payees

Comment (krebernisak) [reply packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/proposeConfig.ts:15]: You can try to adapt this Cosmos vanity addr generator: https://github.com/hukkin/cosmosvanity

File: packages-ts/gauntlet-terra-contracts/src/commands/abstract/executionWrapper.ts [hunk 7]
```diff
@@ -1,53 +1,72 @@
...
+  context: ExecutionContext,
+  inputContext: InputContext<Input, ContractInput>,
 ) => (response: Result<TransactionResponse>) => Promise<any>
 
+export interface Validation<Input> {
+  id: string
+  msgSuccess: string
+  msgFail: string
+  validate: (context: ExecutionContext) => (input: Input) => Promise<boolean>
+}
```
[the remark this answers is not in the document]
Comment (krebernisak) [reply packages-ts/gauntlet-terra-contracts/src/commands/abstract/executionWrapper.ts:35]: Also to consider, `msgSuccess` & `msgFail` are defined on obj creation and are static, why? Wouldn't it be better for the `validate` fn to construct and return those messages dynamically so they can contain data either from the input or fetched by the `validate` fn itself?

[the remark this answers is not in the document]
Comment (krebernisak) [reply packages-ts/gauntlet-terra-contracts/src/commands/abstract/executionWrapper.ts:35]: Ultimately, the validation can just be a simple function type, and it shouldn't know about its own `id`:

```
type Validation<Context, Input> = (ctx: Context, in: Input) => Promise<ValidationResult>;
```

```
const validator1 = (_, n) => n < 10 ? asResult(n) : asError(`BAD NUMBER: ${n} < 10`);
const validator2 = ...;
const validator3 = ...;
const composedValidator = allOf([["A", validator1], ["B", validator2], ["C", validator3]]);
// skip "B" and validate
composedValidator.skip("B")(ctx, input)
```

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/proposeConfig.ts (hunk 6 above)
Comment (RodrigoAD) [reply packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/close/proposeConfig.ts]: https://github.com/smartcontractkit/chainlink-terra/pull/239/files#diff-07ced786da08b434e12acebf624b6588e9307b71be6574b4bcf479d746b892c7R45

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/withdrawFunds.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    ? ((await context.provider.wasm.contractQuery(context.contract, 'link_available_for_payment' as any)) as any).amount
+    : input.amount
+  const recipient = input.recipient || context.wallet.key.accAddress
+  return {
+    amount: new BN(amount).toString(),
+    recipient,
+  }
+}
+
+const validateInput = (input: CommandInput): boolean => {
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/withdrawFunds.ts:42]: Why still use this and not split it into separate validation? Even if just one validation fn (no splitting), I'd still rather we be consistent and use a single API (new one).

File: packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/withdrawFunds.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    amount: flags.amount,
+    recipient: flags.recipient,
+    all: flags.all,
+  }
+}
+
+const makeContractInput = async (input: CommandInput, context: ExecutionContext): Promise<ContractInput> => {
+  const amount = input.all
+    ? ((await context.provider.wasm.contractQuery(context.contract, 'link_available_for_payment' as any)) as any).amount
+    : input.amount
```
Comment (krebernisak) [inline packages-ts/gauntlet-terra-contracts/src/commands/contracts/ocr2/withdrawFunds.ts:34]:
```suggestion
  cosnt queryAvailableAmount = async () =>  ...
  const amount = input.all ? await queryAvailableAmount() : input.amount
```

**Review by krebernisak — APPROVED** [review]
Nice cleanup/improvement! LGTM!

**Pull request merged**